Windows separator fix

joeyespo opened the pull request:
Replaced the in-place separator string with a variable that gets set similar to the way Node does it.

When upgrading Node to a version that includes [this commit](https://github.com/joyent/node/commit/4bd54dad33722e93bae5a16f6b274d08944c24cc), the declaration of `sep` can easily be replaced with the following change to [line 5](https://github.com/sstephenson/stitch/blob/master/src/stitch.coffee#L5).

``` coffee
{extname, join, normalize, sep} = require 'path'
```

joeyespo committed:
Windows separator fix.

robofish commented:
you'd have to adjust line 118 as well:

``` javascript
result += JSON.stringify name.replace /\\/g,'/'
```

plus: one could use the normalize function to get the separator:

``` javascript
sep = normalize '/'
```

joeyespo committed:
Fixed compileSources for Windows compatibility.

joeyespo commented:
@robofish Just addressed your first concern. It's now part of this pull request.

As for `normalize`, yeah you could do that. I personally don't know what additional implications that would have, so I played it safe and just used [the same method Node uses](https://github.com/joyent/node/blob/master/lib/path.js#L23).

robofish commented:
cool :)

Yeah, I'm wasn't too sure about <code>normalize</code> either, but if felt somewhat right to add some abstraction there - on the other hand path separators won't be subject of frequent changes ;)

robofish commented on src/stitch.coffee:
```diff
@@ -113,7 +116,7 @@ exports.Package = class Package
       for name, {filename, source} of sources
         result += if index++ is 0 then "" else ", "
-        result += JSON.stringify name
+        result += if isWindows then JSON.stringify(name.replace(/\\/g, '/')) else name
```
uh wait - you'll need another JSON.stringify for the posix case here

joeyespo replied:
Is that really what we want to do here? When on Unix, the path separators will already correctly be '/'.

robofish replied:
yeah, but the the point of this change is windows compatability, right? ;)
and I do have first hand knowledge that the packages will have <code>\\</code> in their path name, while the require statements use / as path separator => require won't find the package.
my point was that your latest change removes the 'JSON.stringify' from the name (which basically just adds the quotes) in case it's an unix system. you might want to do something like <code>JSON.stringify( if isWindows then name.replace(/\\/g, '/') else name )</code>

joeyespo replied:
> and I do have first hand knowledge that the packages will have \ in their path name

Ah, ok. I actually think this is a mistake.

"Windows compatibility" means you should be able to run Node on Windows, install the Stitch module, and everything should still work. It does _not_ mean, for example, `require` paths should suddenly start using '\' in their paths. This would tie your code to the system it's developed on, making collaboration nearly impossible. Requiring that stitch uses Unix paths parallels the uniformity of URL paths.

That said, if Node accepts both forward and back slashes for `require` for paths, stitch should also. Though it should accept both forms from all operating systems like Node and not show different error messages that depend on what OS Node happens to be running on.

joeyespo replied:
Just updated the pull request with better path normalization. Also, both path styles can now be used in `require` like with Node. See 2e4e6cdb2be204740ca341da787aaac00d347721.

joeyespo committed:
Windows support fixes.

* Paths are now properly normalized upfront so both Windows-style and Unix-style paths are supported to behave like Node's require
* Module names are now always Unix-style, even when Windows-style names are used when calling "require"
* Consequently, error messages are always shown in Unix-style, which will encourage OS-independent code
  e.g. "require('lib\\missing')" displays "'lib/missing' not found" regardless of the underlying OS

spikebrehm commented:
@sstephenson do you think you'll be able to merge this in soon?

HenrikJoreteg commented:
+1
